File: libraries/cyclestreets-view/src/main/java/net/cyclestreets/routing/Segments.java
package net.cyclestreets.routing;

import java.util.Iterator;
import java.util.LinkedList;

import net.cyclestreets.util.Turn;
import org.osmdroid.api.IGeoPoint;

public class Segments implements Iterable<Segment>
{
  private final LinkedList<Segment> segments = new LinkedList<>();

  public Segments() {}

  public int count() { return segments.size(); }
  public boolean isEmpty() { return segments.isEmpty(); }

  public IGeoPoint startPoint() { return segments.getFirst().start(); }
  public IGeoPoint finishPoint() { return segments.getLast().finish(); }
  public Segment.Start first() { return (Segment.Start) segments.getFirst(); }
  public Segment.End last() { return (Segment.End) segments.getLast(); }

  public void add(final Segment seg) {
    if (seg instanceof Segment.Start) {
      segments.addFirst(seg);
      return;
    }

    if (count() != 0) {
      final Segment previous = segments.get(count() - 1);

      // Meld "Join Roundabout" instructions
      if (Turn.JOIN_ROUNDABOUT.equals(previous.turn)) {
        segments.remove(previous);
        segments.add(new Segment.Step(previous, seg, seg.turn, seg.turnInstruction));
        return;
      }

      // Meld staggered crossroads
      if (previous.distance_ < 20) {
        if (Turn.TURN_LEFT.equals(previous.turn) && Turn.TURN_RIGHT.equals(seg.turn)) {
          segments.remove(previous);
          segments.add(new Segment.Step(previous, seg, Turn.LEFT_RIGHT, Turn.LEFT_RIGHT.getTextInstruction()));
          return;
        }
        if (Turn.TURN_RIGHT.equals(previous.turn) && Turn.TURN_LEFT.equals(seg.turn)) {
          segments.remove(previous);
          segments.add(new Segment.Step(previous, seg, Turn.RIGHT_LEFT, Turn.RIGHT_LEFT.getTextInstruction()));
          return;
        }
      }

      // Meld bridges
      if (previous.distance_ < 100 && "Bridge".equalsIgnoreCase(previous.name_) && seg.turn == Turn.STRAIGHT_ON) {
        segments.remove(previous);
        segments.add(new Segment.Step(previous, seg, previous.turn, previous.turnInstruction + " over Bridge"));
        return;
      }
    }

    segments.add(seg);
  }

  public Segment get(final int i) { return segments.get(i); }

  @Override
  public Iterator<Segment> iterator() { return segments.iterator(); }

  public Iterator<IGeoPoint> pointsIterator() { return new PointsIterator(this); }

  private static class PointsIterator implements Iterator<IGeoPoint>  {
    private final Iterator<Segment> segments_;
    private Iterator<IGeoPoint> points_;

    PointsIterator(final Segments segments) {
      segments_ = segments.iterator();
      if (!segments_.hasNext())
        return;

      points_ = segments_.next().points();
    }

    @Override
    public boolean hasNext() {
      return points_ != null && points_.hasNext();
    }

    @Override
    public IGeoPoint next() {
      if (!hasNext())
        throw new IllegalStateException();

      final IGeoPoint p = points_.next();

      if (!hasNext()) {
        if (segments_.hasNext())
          points_ = segments_.next().points();
        else
          points_ = null;
      }

      return p;
    }

    @Override
    public void remove() {
      throw new UnsupportedOperationException();
    }
  }

}

File: libraries/cyclestreets-view/src/main/java/net/cyclestreets/routing/Segments.kt
package net.cyclestreets.routing

import java.util.LinkedList

import net.cyclestreets.util.Turn
import net.cyclestreets.util.Turn.*
import org.osmdroid.api.IGeoPoint

class Segments : Iterable<Segment> {
    private val segments = LinkedList<Segment>()

    fun count(): Int { return segments.size }
    fun isEmpty(): Boolean { return segments.isEmpty() }

    fun startPoint(): IGeoPoint { return segments.first.start() }
    fun finishPoint(): IGeoPoint { return segments.last.finish() }
    fun first(): Segment.Start { return segments.first as Segment.Start }
    fun last(): Segment.End { return segments.last as Segment.End }

    private val CROSSROAD_MELDS: Map<Pair<Turn, Turn>, Turn> = mapOf(
        Pair(TURN_LEFT, TURN_RIGHT) to TURN_LEFT_THEN_RIGHT,
        Pair(TURN_RIGHT, TURN_LEFT) to TURN_RIGHT_THEN_LEFT,
        Pair(BEAR_LEFT, BEAR_RIGHT) to BEAR_LEFT_THEN_RIGHT,
        Pair(BEAR_RIGHT, BEAR_LEFT) to BEAR_RIGHT_THEN_LEFT
    )

    fun add(seg: Segment) {
        if (seg is Segment.Start) {
            segments.addFirst(seg)
            return
        }

        if (count() != 0) {
            val previous = segments[count() - 1]

            // Meld "Join Roundabout" instructions
            if (Turn.JOIN_ROUNDABOUT == previous.turn) {
                segments.remove(previous)
                segments.add(Segment.Step(previous, seg, seg.turn, seg.turnInstruction))
                return
            }

            // Meld staggered crossroads
            if (previous.distance < 20) {
                CROSSROAD_MELDS[Pair(previous.turn, seg.turn)]?.let {
                    segments.remove(previous)
                    segments.add(Segment.Step(previous, seg, it, it.textInstruction))
                    return
                }
            }

            // Meld bridges
            if (previous.distance < 100 && "Bridge".equals(previous.name, ignoreCase = true) && Turn.STRAIGHT_ON === seg.turn) {
                segments.remove(previous)
                segments.add(Segment.Step(previous, seg, previous.turn, previous.turnInstruction + " over Bridge"))
                return
            }
        }

        segments.add(seg)
    }

    operator fun get(i: Int): Segment { return segments[i] }

    override fun iterator(): Iterator<Segment> { return segments.iterator() }
    fun pointsIterator(): Iterator<IGeoPoint> { return PointsIterator(this) }

    private class PointsIterator internal constructor(segments: Segments) : Iterator<IGeoPoint> {
        private val segmentIterator: Iterator<Segment> = segments.iterator()
        private var pointIterator: Iterator<IGeoPoint>? = null

        init {
            if (segmentIterator.hasNext())
                pointIterator = segmentIterator.next().points()
        }

        override fun hasNext(): Boolean {
            return pointIterator?.hasNext() ?: false
        }

        override fun next(): IGeoPoint {
            if (!hasNext())
                throw IllegalStateException()

            val p = pointIterator!!.next()

            if (!hasNext()) {
                if (segmentIterator.hasNext())
                    pointIterator = segmentIterator.next().points()
                else
                    pointIterator = null
            }

            return p
        }
    }
}
